File: Data Structures & Algorithms/two-integer-sum/submission-1.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashMap = dict()

        for i in range(len(nums)):
            hashMap[nums[i]] = i

        for i in range(len(nums)):
            difference = target - nums[i]

            if difference in hashMap and hashMap.get(difference) != i:
                return [i, hashMap.get(difference)]
        

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // edge cases
        if (nums.size() == 1) return nums[0] == target ? 0 : -1;

        int left = 0; 
        int right = nums.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) return mid;

            // if in Right Sorted
            if (nums[mid] >= nums[left]) {
                // is target within this half?
                if (target >= nums[left] && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                // is target within this half?
                if (target > nums[mid] && target <= nums[right]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }

        return -1;
    }
};
